Cleaned up infantry rules (part 1)

reaperrr opened the pull request:
This is a first step to reduce the large amount of (often unnecessary) duplication in yaml rules as well as a sort-of a prerequisite for #8012 (would reduce the required yaml changes in that PR greatly).

TS:
- Introduces a `^Viceroid` default to TS, to cut down removals/duplications
- Sets default `Bounds` on infantry default to reduce duplication, fixes position
- Moves `TakeCover` definitions to infantry default to cut down yaml lines

D2k:
- Sets default `Bounds` on infantry default to reduce duplication, fixes position
- Removes `TakeCover` properties from individual infantry types to reduce duplication
- Adds prone sequences for engineer and medic/thumper.

TD:
- Sets default `Bounds` on infantry default to reduce duplication
- Adds `TakeCover` trait to individual infantry to work around issue that prone anims don't show

RA:
- Sets default `Bounds` on infantry default to reduce duplication
- Moves `TakeCover` properties to infantry default to reduce duplication

reaperrr committed:
Cleaned and fixed TS infantry rules a little.

Moved Selectable Bounds to infantry default and tweaked them to better
match art size.
Gave Viceroids their own default. They don't share cells anymore, either.
Fixed JumJet voice.
Disabled RevealsShroud on Tiberian Fiend.
Cleaned up TakeCover definitions.
Tweaked selection bounds of TS cyborgs.
Changed TS engineer trait order to fix prone animations not showing up.

reaperrr committed:
Cleaned and fixed TD infantry rules a little

Fixes TD infantry not going visually prone.

reaperrr committed:
Cleaned and fixed D2k infantry rules a little

Moved Selectable Bounds to infantry default, fixed vertical offset.
Removed some TakeCover duplication.
Added engineer and medic/thumper prone sequences.
Made sure that TakeCover/WithInfantryBody sequence modifiers are applied
properly.

Fixed d2k grenadier sequences offset to match other infantry.

reaperrr committed:
Cleaned RA infantry rules a little

abcdefg30 commented:
> TD:
•Moves  TakeCover  to infantry default to reduce duplication

I don't see this change and TD infantry doesn't take cover for me.

reaperrr commented:
That was just a copy-paste error in the description.

That being said, the default inf definition has that trait so if it doesn't take cover, something is wrong here (on bleed, too).

penev92 commented:
That would be on me.

abcdefg30 commented:
Take cover did not work for me in ra as well.

penev92 commented:
Confirmed that it does indeed work. Refer to [this comment](https://github.com/OpenRA/OpenRA/pull/8011#issuecomment-96334245) for an explanation.

reaperrr commented:
> Confirmed that it does indeed work. Refer to this comment for an explanation.

Okay. It seems to have gotten worse since #8011 was merged though, so we should fix that graphical bug before next release.

As for this PR, I forgot to do the TakeCover cleanup for TS, so even if the rest is fine, don't merge this before I've amended that.

abcdefg30 commented:
:+1: / :white_check_mark: 
when following two things are fixed:
- [ ] "As for this PR, I forgot to do the TakeCover cleanup for TS, so even if the rest is fine, don't merge this before I've amended that."
- [ ] The viceroid selection bounds are wrong

reaperrr commented:
Fixing #8075 is out of the scope of this PR, so there's 2 ways to tackle the TakeCover issue:
1) Remove the TakeCover-related changes from this PR and re-introduce them later together with a fix for #8075 (clean solution)
2) Only move the DamageTriggers and DamageModifiers to the infantry defaults and leave ```TakeCover:``` in place for each actor to work around #8075 (somewhat hacky solution).

Personally, I favor 1). Other opinions?

pchote commented:
#8075 is a simple < 5 line fix, so I suggest:

3) Create a new pr that fixes #8075, and then make this depend on that.

penev92 commented:
Like I said on #8075, I tried fixing it, but the problem isn't only one. We'll need to either 1. change how sequence modifiers work (provide sequence name instead of prefix), 2. add the missing sequences, or 3. make `WithInfantryBody` keep trying until it gets a valid sequence.

Personally I vote for 1.

reaperrr commented:
I've decided to give #8012 priority, I'll return to this later.

reaperrr commented:
Rebased & updated, added a yaml-level workaround for prone anims not showing in TD.

Note: This is mostly about the Selectable clean-up, the rest is a 'bonus'. If the TakeCover work-around isn't wanted, I'd rather remove it from this PR entirely so we can get the Selectable cleanup merged, I need it for an upcoming PR.

penev92 commented on mods/d2k/rules/infantry.yaml:
```diff
@@ -49,6 +42,7 @@ engineer:
 	Captures:
 		CaptureTypes: husk
 	-AutoTarget:
+	-TakeCover:
```
Why is this?

reaperrr replied:
The engineer has no prone sequences.

penev92 replied:
The AssetBrowser suggests otherwise, if I'm not mistaking. Likely the sequence just isn't defined, in which case better leave a note here saying that the engineer is missing his sequence.

reaperrr replied:
I have an even better idea, I'll take a look at the asset browser and just add those sequences ;)

penev92 replied:
That would be great ;)

reaperrr commented:
Rebased, updated, squashed TD commits, reordered commits.

Compared to the last update, I amended the TakeCover cleanup for TS, added prone sequences for d2k engineer & thumper/medic and made sure each inf in d2k has both TakeCover and WithInfantryBody on the same yaml 'level' to work around the sequence modifier issues until we fix that at code level.

penev92 commented on mods/cnc/rules/defaults.yaml:
```diff
@@ -180,6 +180,7 @@
 			Beach: 80
 	SelectionDecorations:
 	Selectable:
+		Bounds: 12,17,0,-6
```
Tiny nit: RA and D2k use `12,18,0,-6`.
I guess TS has bigger infantry, but these use the same size, so you can use the same bounds for consistency.

reaperrr replied:
No, TD infantry is smaller. The difference is more than 1 pixel actually, but I didn't feel like scaling 100% accurately because otherwise either TD Bounds would end up too small (making it harder to select them) or RA/D2k bounds too large (risking overlapping)

penev92 commented:
OK, now all mods have both `TakeCover` and `WithInfantryBody` on both the `^Infantry` default and on each soldier type. Is that to work around the same sequence modifier issue? Because it's not pretty.

Changes look good. I still haven't tested ingame.
I still don't think the viceroid default needs `Cloak`, `UpgradeManage`, `DeathSounds` or  `Guard`. Probably doesn't need 'UpdatesPlayerStatistics` and `DrawLineToTarget` either, but those need to be verified.

All in all :+0.5: until I test this.

reaperrr commented:
> OK, now all mods have both TakeCover and WithInfantryBody on both the ^Infantry default and on each soldier type. Is that to work around the same sequence modifier issue? Because it's not pretty.

Yes and I know. That's a short-term solution for the upcoming release (provided this gets merged in time), someone should definitely fix that properly for next+1 so we can get rid of these workarounds entirely.

reaperrr commented:
Updated, removed Voiced, DeathSounds, Cloak, UpgradeManager and Guard from viceroid default.

penev92 commented:
I have no idea if this is related, but the grenadier looks off:
![](http://i.imgur.com/1Q4rvWs.png)

Also I didn't see a problem in D2k, but in RA and TD the workaround only works half of the time :/ (probably also in D2k, and I just got lucky)

The bounds of both the cyborg and the cyborg commando look like they can be moved up a bit:
![](http://i.imgur.com/HzPaJoy.png)

And I couldn't get a single Engineer in TS to take cover.

I'm not sure if some of those things are even fixable here, so here's my :+1: to another very nice cleanup.

reaperrr commented:
Improved cyborg bounds:
![ts-cyb-sel-offsets](https://cloud.githubusercontent.com/assets/2857877/7903437/d9b5a2c6-07dc-11e5-8cb3-3d64d44dc0bf.png)

> I have no idea if this is related, but the grenadier looks off:

The art/sequences needed a different offset. Fixed:
![d2k-gren-offset](https://cloud.githubusercontent.com/assets/2857877/7903433/c58bb542-07dc-11e5-9512-eecaf1e277c1.png)

> And I couldn't get a single Engineer in TS to take cover.

Apparently, even the order(!) of the traits matters, putting TakeCover first fixed this.
We really need to deal with that code as soon as possible...

penev92 commented:
Looks like this needs another rebase.

reaperrr commented:
Rebased.

abcdefg30 commented:
> I have no idea if this is related, but the grenadier looks off:
The art/sequences needed a different offset. Fixed:

This broke the icon:
![grenadiericonoffset](https://cloud.githubusercontent.com/assets/7704140/7918364/1bd005a2-0891-11e5-98f8-6e3ea2d36514.png)
You should add `Offset: 0,0` to the icon's sequence.

abcdefg30 commented:
I also couldn't get the `British spy`'s and ra `Engineer`'s prone animation.

abcdefg30 commented:
Td `Commando` and `Chemical warrior` also don't have prone-stand animations for me.

penev92 commented:
They did for me. Looks like there is some race condition and like I said - it works about half of the time. :/

reaperrr commented:
Updated, fixed the saboteur icon and added ```TakeCover:``` to british spy. The other cases must be the work of the race condition :/

Mailaender commented:
Thanks. Couldn't spot regressions. :white_check_mark: [Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5440cf787923c035ccb27e12441af7397e8d522b...3eaa02ccddfc6db0d27f49aba32e2e3c7abfa2c5)
